### PR TITLE
Don't add redundant help for object safety violations

### LINKED ISSUE #117186: Don't show the same help multiple times for object safety violations
#116401 made it so that we report all object safety violations at the same time. However, it also results in a redundant note:

https://github.com/rust-lang/rust/pull/116401/files#diff-ea89a7ba6273e575f66679f7c4a10297572773dc805cc27eb6398f7c12527ab8R16

We should not add the same note multiple times.

### PR DESCRIPTION
**Pull request by rmehri01**
Fixes #117186

r? WaffleLapkin

### CODE REVIEW

File: tests/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr [hunk 1]
```diff
@@ -62,7 +62,6 @@ LL | trait NonObjectSafe4 {
    |       -------------- this trait cannot be made into an object...
 LL |     fn foo(&self, s: &Self);
    |                      ^^^^^ ...because method `foo` references the `Self` type in this parameter
-   = help: consider moving `foo` to another trait
```
Comment (WaffleLapkin) [inline tests/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr:65]: I don't think this is a change we want. Note that this mentions a different `foo` than the previous messages.

Comment (rmehri01) [reply tests/ui/feature-gates/feature-gate-object_safe_for_dispatch.stderr]: Ah okay, sorry about that, I pushed a different solution that doesn't change the other tests

**Comment by spastorino** [comment]
@WaffleLapkin are you taking care of this review? can you assign it to yourselve?.
My only thought from skimming it quickly is ... why are we reporting twice in the first place?, do we have an idea why we insert 2 times in this list?.

**Comment by WaffleLapkin** [comment]
@spastorino I can take the review.

The cause is that there can be multiple problems with the same item, we want to highlight all of them, but the proposed solution is the same for all problems — move the item to another trait.

**Comment by spastorino** [comment]
> The cause is that there can be multiple problems with the same item, we want to highlight all of them, but the proposed solution is the same for all problems — move the item to another trait.

Makes sense, I guess I'd hide the existence of that vector that tracks already suggested "solutions".

File: compiler/rustc_infer/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -103,9 +103,11 @@ pub fn report_object_safety_error<'tcx>(
     if trait_span.is_some() {
         let mut reported_violations: Vec<_> = reported_violations.into_iter().collect();
         reported_violations.sort();
+
+        let mut reported_names = vec![];
```
Comment (WaffleLapkin) [inline compiler/rustc_infer/src/traits/error_reporting/mod.rs:107]: Could you add a little comment explaining why this is needed like "allows us to skip suggesting moving the same item to another trait multiple times" or something?

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
This looks fine.

Please add a comment, fix [conflicts](https://rustc-dev-guide.rust-lang.org/git.html?highlight=confl#rebasing-and-conflicts) and use `@rustbot review` to ask me for review again =)

File: compiler/rustc_infer/src/traits/error_reporting/mod.rs (hunk 2 above)
Comment (rmehri01) [reply compiler/rustc_infer/src/traits/error_reporting/mod.rs:109]: Done! Do you think there is a better way of doing this to hide the vector, maybe by taking in the list of violations up front for example or is it okay as is?

Comment (WaffleLapkin) [reply compiler/rustc_infer/src/traits/error_reporting/mod.rs:109]: We could create an `ObjectSafetyViolationSolution` enum, which would be returned from `solution`, then deduplicated, then added to the error. Something like
```rust
let mut potential_solutions: Vec<_> = reported_violations.into_iter().map(...::solution).collect();
potential_solutions.sort();
potential_solutions.dedup();
potential_solutions.for_each(|solution| solution.add_to(&mut err));
```

Do you want to try doing this?

Comment (rmehri01) [reply compiler/rustc_infer/src/traits/error_reporting/mod.rs:109]: Ah I see, that sounds better, I'll give it a try!

File: compiler/rustc_infer/src/traits/error_reporting/mod.rs [hunk 3]
```diff
@@ -104,9 +104,15 @@ pub fn report_object_safety_error<'tcx>(
     if trait_span.is_some() {
         let mut reported_violations: Vec<_> = reported_violations.into_iter().collect();
         reported_violations.sort();
-        for violation in reported_violations {
-            // Only provide the help if its a local trait, otherwise it's not actionable.
-            violation.solution(&mut err);
+
+        // Only provide the help if its a local trait, otherwise it's not actionable.
```
Comment (WaffleLapkin) [inline compiler/rustc_infer/src/traits/error_reporting/mod.rs:108]: Please move this to the `if`

**Review by WaffleLapkin — APPROVED** [review]
Thanks, I like this fix :)

**Pull request merged**